### ISSUE TITLE
Problem installing 3.0.5 on OS X

### ISSUE THREAD

**Issue by PeterWAWood** [ruby-gnome2/ruby-gnome2#584]
I automatically upgraded to 3.0.5 on OS X using gem upgrade. The native extensions for gobject-introspection and gtk3 did not compile (as I use Homebrew).

I then installed gtk3 and gobject-introspection by setting the PKG_CONFIG_PATH:

PKG_CONFIG_PATH=/usr/local/opt/libffi/lib/pkgconfig gem install gtk3

When I tried a simple Hello Wold script, I got this error
```
mbp:gtk3 peter$ ruby hello_world.rb
/usr/local/lib/ruby/gems/2.2.0/gems/gobject-introspection-3.0.5/lib/gobject-introspection.rb:34:in `<top (required)>': uninitialized constant Pathname (NameError)
	from /usr/local/Cellar/ruby/2.2.3/lib/ruby/2.2.0/rubygems/core_ext/kernel_require.rb:69:in `require'
	from /usr/local/Cellar/ruby/2.2.3/lib/ruby/2.2.0/rubygems/core_ext/kernel_require.rb:69:in `require'
	from /usr/local/lib/ruby/gems/2.2.0/gems/gtk3-3.0.5/lib/gtk3.rb:17:in `<top (required)>'
	from /usr/local/Cellar/ruby/2.2.3/lib/ruby/2.2.0/rubygems/core_ext/kernel_require.rb:128:in `require'
	from /usr/local/Cellar/ruby/2.2.3/lib/ruby/2.2.0/rubygems/core_ext/kernel_require.rb:128:in `rescue in require'
	from /usr/local/Cellar/ruby/2.2.3/lib/ruby/2.2.0/rubygems/core_ext/kernel_require.rb:39:in `require'
	from hello_world.rb:1:in `<main>'
```

I'm guessing that I haven't installed gobject-introspection (or one of its dependencies) properly.

**Comment by kou** [comment]
Could you show output of the following commands?

```text
% ruby -r glib2 -e 'puts $:'
```

**Comment by PeterWAWood** [comment]
Here it is:
```
mbp:~ peter$ ruby -r glib2 -e 'puts $:'
/usr/local/lib/ruby/gems/2.2.0/gems/pkg-config-1.1.6/lib
/usr/local/lib/ruby/gems/2.2.0/extensions/x86_64-darwin-14/2.2.0/glib2-3.0.5
/usr/local/lib/ruby/gems/2.2.0/gems/glib2-3.0.5/lib
/usr/local/lib/ruby/site_ruby/2.2.0
/usr/local/lib/ruby/site_ruby/2.2.0/x86_64-darwin14
/usr/local/lib/ruby/site_ruby
/usr/local/lib/ruby/vendor_ruby/2.2.0
/usr/local/lib/ruby/vendor_ruby/2.2.0/x86_64-darwin14
/usr/local/lib/ruby/vendor_ruby
/usr/local/Cellar/ruby/2.2.3/lib/ruby/2.2.0
/usr/local/Cellar/ruby/2.2.3/lib/ruby/2.2.0/x86_64-darwin14
```

**Comment by kou** [comment]
Oh... Sorry. I write wrong command line...

Could you try the following again?

```text
% ruby -r glib2 -e 'puts $"'
```

**Comment by PeterWAWood** [comment]
No problem  - thanks for looking into my issue:
```
mbp:~ peter$ ruby -r glib2 -e 'puts $"'
enumerator.so
rational.so
complex.so
/usr/local/Cellar/ruby/2.2.3/lib/ruby/2.2.0/x86_64-darwin14/enc/encdb.bundle
/usr/local/Cellar/ruby/2.2.3/lib/ruby/2.2.0/x86_64-darwin14/enc/trans/transdb.bundle
/usr/local/Cellar/ruby/2.2.3/lib/ruby/2.2.0/unicode_normalize.rb
/usr/local/Cellar/ruby/2.2.3/lib/ruby/2.2.0/x86_64-darwin14/rbconfig.rb
thread.rb
/usr/local/Cellar/ruby/2.2.3/lib/ruby/2.2.0/x86_64-darwin14/thread.bundle
/usr/local/Cellar/ruby/2.2.3/lib/ruby/2.2.0/rubygems/compatibility.rb
/usr/local/Cellar/ruby/2.2.3/lib/ruby/2.2.0/rubygems/defaults.rb
/usr/local/Cellar/ruby/2.2.3/lib/ruby/2.2.0/rubygems/deprecate.rb
/usr/local/Cellar/ruby/2.2.3/lib/ruby/2.2.0/rubygems/errors.rb
/usr/local/Cellar/ruby/2.2.3/lib/ruby/2.2.0/rubygems/version.rb
/usr/local/Cellar/ruby/2.2.3/lib/ruby/2.2.0/rubygems/requirement.rb
/usr/local/Cellar/ruby/2.2.3/lib/ruby/2.2.0/rubygems/platform.rb
/usr/local/Cellar/ruby/2.2.3/lib/ruby/2.2.0/rubygems/basic_specification.rb
/usr/local/Cellar/ruby/2.2.3/lib/ruby/2.2.0/rubygems/stub_specification.rb
/usr/local/Cellar/ruby/2.2.3/lib/ruby/2.2.0/rubygems/util/stringio.rb
/usr/local/Cellar/ruby/2.2.3/lib/ruby/2.2.0/rubygems/specification.rb
/usr/local/Cellar/ruby/2.2.3/lib/ruby/2.2.0/rubygems/exceptions.rb
/usr/local/Cellar/ruby/2.2.3/lib/ruby/2.2.0/rubygems/defaults/operating_system.rb
/usr/local/Cellar/ruby/2.2.3/lib/ruby/2.2.0/rubygems/core_ext/kernel_gem.rb
/usr/local/Cellar/ruby/2.2.3/lib/ruby/2.2.0/monitor.rb
/usr/local/Cellar/ruby/2.2.3/lib/ruby/2.2.0/rubygems/core_ext/kernel_require.rb
/usr/local/Cellar/ruby/2.2.3/lib/ruby/2.2.0/rubygems.rb
/usr/local/Cellar/ruby/2.2.3/lib/ruby/2.2.0/rubygems/path_support.rb
/usr/local/Cellar/ruby/2.2.3/lib/ruby/2.2.0/rubygems/dependency.rb
/usr/local/Cellar/ruby/2.2.3/lib/ruby/2.2.0/x86_64-darwin14/pathname.bundle
/usr/local/Cellar/ruby/2.2.3/lib/ruby/2.2.0/pathname.rb
/usr/local/Cellar/ruby/2.2.3/lib/ruby/2.2.0/English.rb
/usr/local/lib/ruby/gems/2.2.0/gems/glib2-3.0.5/lib/glib2/deprecatable.rb
/usr/local/lib/ruby/gems/2.2.0/extensions/x86_64-darwin-14/2.2.0/glib2-3.0.5/glib2.bundle
/usr/local/lib/ruby/gems/2.2.0/gems/glib2-3.0.5/lib/glib2.rb
```

**Comment by kou** [comment]
Thanks.

It's strange that `ininitialized constant Pathname` is occurred because `pathname` is required:

```text
/usr/local/Cellar/ruby/2.2.3/lib/ruby/2.2.0/x86_64-darwin14/pathname.bundle
/usr/local/Cellar/ruby/2.2.3/lib/ruby/2.2.0/pathname.rb
```

Could you add `puts $"` line before the following line at /usr/local/lib/ruby/gems/2.2.0/gems/gobject-introspection-3.0.5/lib/gobject-introspection.rb:34 and show the output?

```ruby
base_dir = Pathname.new(__FILE__).dirname.dirname.expand_path
```

**Comment by PeterWAWood** [comment]
If I did it correctly, this is the output:
```
mbp:gtk3 peter$ ruby hello_world.rb
enumerator.so
rational.so
complex.so
/usr/local/Cellar/ruby/2.2.3/lib/ruby/2.2.0/x86_64-darwin14/enc/encdb.bundle
/usr/local/Cellar/ruby/2.2.3/lib/ruby/2.2.0/x86_64-darwin14/enc/trans/transdb.bundle
/usr/local/Cellar/ruby/2.2.3/lib/ruby/2.2.0/unicode_normalize.rb
/usr/local/Cellar/ruby/2.2.3/lib/ruby/2.2.0/x86_64-darwin14/rbconfig.rb
thread.rb
/usr/local/Cellar/ruby/2.2.3/lib/ruby/2.2.0/x86_64-darwin14/thread.bundle
/usr/local/Cellar/ruby/2.2.3/lib/ruby/2.2.0/rubygems/compatibility.rb
/usr/local/Cellar/ruby/2.2.3/lib/ruby/2.2.0/rubygems/defaults.rb
/usr/local/Cellar/ruby/2.2.3/lib/ruby/2.2.0/rubygems/deprecate.rb
/usr/local/Cellar/ruby/2.2.3/lib/ruby/2.2.0/rubygems/errors.rb
/usr/local/Cellar/ruby/2.2.3/lib/ruby/2.2.0/rubygems/version.rb
/usr/local/Cellar/ruby/2.2.3/lib/ruby/2.2.0/rubygems/requirement.rb
/usr/local/Cellar/ruby/2.2.3/lib/ruby/2.2.0/rubygems/platform.rb
/usr/local/Cellar/ruby/2.2.3/lib/ruby/2.2.0/rubygems/basic_specification.rb
/usr/local/Cellar/ruby/2.2.3/lib/ruby/2.2.0/rubygems/stub_specification.rb
/usr/local/Cellar/ruby/2.2.3/lib/ruby/2.2.0/rubygems/util/stringio.rb
/usr/local/Cellar/ruby/2.2.3/lib/ruby/2.2.0/rubygems/specification.rb
/usr/local/Cellar/ruby/2.2.3/lib/ruby/2.2.0/rubygems/exceptions.rb
/usr/local/Cellar/ruby/2.2.3/lib/ruby/2.2.0/rubygems/defaults/operating_system.rb
/usr/local/Cellar/ruby/2.2.3/lib/ruby/2.2.0/rubygems/core_ext/kernel_gem.rb
/usr/local/Cellar/ruby/2.2.3/lib/ruby/2.2.0/monitor.rb
/usr/local/Cellar/ruby/2.2.3/lib/ruby/2.2.0/rubygems/core_ext/kernel_require.rb
/usr/local/Cellar/ruby/2.2.3/lib/ruby/2.2.0/rubygems.rb
/usr/local/Cellar/ruby/2.2.3/lib/ruby/2.2.0/rubygems/path_support.rb
/usr/local/Cellar/ruby/2.2.3/lib/ruby/2.2.0/rubygems/dependency.rb
/usr/local/lib/ruby/gems/2.2.0/extensions/x86_64-darwin-14/2.2.0/glib2-3.0.5/glib2.bundle
/usr/local/lib/ruby/gems/2.2.0/gems/gobject-introspection-3.0.5/lib/gobject-introspection.rb:35:in `<top (required)>': uninitialized constant Pathname (NameError)
	from /usr/local/Cellar/ruby/2.2.3/lib/ruby/2.2.0/rubygems/core_ext/kernel_require.rb:69:in `require'
	from /usr/local/Cellar/ruby/2.2.3/lib/ruby/2.2.0/rubygems/core_ext/kernel_require.rb:69:in `require'
	from /usr/local/lib/ruby/gems/2.2.0/gems/gtk3-3.0.5/lib/gtk3.rb:17:in `<top (required)>'
	from /usr/local/Cellar/ruby/2.2.3/lib/ruby/2.2.0/rubygems/core_ext/kernel_require.rb:128:in `require'
	from /usr/local/Cellar/ruby/2.2.3/lib/ruby/2.2.0/rubygems/core_ext/kernel_require.rb:128:in `rescue in require'
	from /usr/local/Cellar/ruby/2.2.3/lib/ruby/2.2.0/rubygems/core_ext/kernel_require.rb:39:in `require'
	from hello_world.rb:1:in `<main>'
```

**Comment by kou** [comment]
It's strange that `/usr/local/lib/ruby/gems/2.2.0/gems/glib2-3.0.5/lib/glib2.rb` isn't required. Could you try to look into this?

**Comment by PeterWAWood** [comment]
I added ```require 'glib2'``` to my test script and it runs fine. I try and take a look into why it didn't get required automatically.

**Comment by kou** [comment]
Thanks!

**Comment by PeterWAWood** [comment]
I've done a little simple debugging and found something that seems odd.

I added a simple puts statement at the front of gtk3.rb, gobject-introspection.rb and glib2.rb to print that the file was loading. I also added a puts statement to print the contents of base_dir in gobject-introspection.rb. When I require gtk3, glib2 does not seem to load. If I directly require gobjet-introspection, glib2 is loaded.

Here is what happens in irb: 
```
mbp:gtk3 peter$ irb 
irb(main):001:0> require 'gtk3'
gobject-introspection being loaded
NameError: uninitialized constant Pathname
	from /usr/local/lib/ruby/gems/2.2.0/gems/gobject-introspection-3.0.6/lib/gobject-introspection.rb:35:in `<top (required)>'
	from /usr/local/Cellar/ruby/2.2.3/lib/ruby/2.2.0/rubygems/core_ext/kernel_require.rb:69:in `require'
	from /usr/local/Cellar/ruby/2.2.3/lib/ruby/2.2.0/rubygems/core_ext/kernel_require.rb:69:in `require'
	from /usr/local/lib/ruby/gems/2.2.0/gems/gtk3-3.0.6/lib/gtk3.rb:17:in `<top (required)>'
	from /usr/local/Cellar/ruby/2.2.3/lib/ruby/2.2.0/rubygems/core_ext/kernel_require.rb:128:in `require'
	from /usr/local/Cellar/ruby/2.2.3/lib/ruby/2.2.0/rubygems/core_ext/kernel_require.rb:128:in `rescue in require'
	from /usr/local/Cellar/ruby/2.2.3/lib/ruby/2.2.0/rubygems/core_ext/kernel_require.rb:39:in `require'
	from (irb):1
	from /usr/local/bin/irb:11:in `<main>'
irb(main):002:0> quit
```
```
mbp:gtk3 peter$ irb
irb(main):001:0> require 'gobject-introspection'
gobject-introspection being loaded
glib2 being loaded
/usr/local/lib/ruby/gems/2.2.0/gems/gobject-introspection-3.0.6
=> true
irb(main):002:0> 
```
Any suggestions as to what I can do to find out why this is happening?

(I have updated to 3.0.6)

**Comment by kou** [comment]
Could OS X users help us?

**Comment by lemsx1** [comment]
I have this same problem with El Capitan and i simply edited pkg-config.rb with:

      @@paths = [ '/usr/local/opt/libffi/lib/pkgconfig' ]

**Comment by cosmo0920** [comment]
I've hit same problem a long ago.
I resolve this problem locally by defining `PKG_CONFIG_PATH` to find `libffi` and `libxml2`.
How about define `PKG_CONFIG_PATH` like this?

```bash
export PKG_CONFIG_PATH=/usr/local/opt/libffi/lib/pkgconfig:/usr/local/opt/libxml2/lib/pkgconfig:$PKG_CONFIG_PATH
```

Then,

```log
% bundle exec irb
irb(main):001:0> require 'gobject-introspection'
=> true
irb(main):002:0> require 'gtk3'
=> true
```

I hope this helps you.

**Comment by PeterWAWood** [comment]
Thanks for your suggestions. I have been able to successfully install gtk3 by supplying the alternative PKG_CONFIG_PATH.

The issue that I'm struggling to identify is the necessity to specifically require 'gobject-introspection' before requiring 'gtk3'. gtk3.rb includes a require statement for 'gobject-introspection' which in turn has a require statement for 'glib2' which in turn requires a number of other modules including pathname.

If you do not specifically require 'gobject-introspection', require 'gtk3' loads 'gobject-introspection' but it does not successfully load 'glib2' or its dependencies. When you specifically require 'gobject-introspection', 'glib2' and its dependencies get properly loaded.

I notice that both of you specifically require 'gobject-introspection' prior to requiring 'gtk3'. As I understand, this shouldn't really be necessary as 'gtk3' has a statement to require 'gobject-introspection'.

Is there a reason that you specifically require 'gobject-introspection' and then require 'gtk3'?

**Comment by mvz** [comment]
There are issues with load path ordering and having both `glib2.so` and `glib2.rb`: In certain cases `require 'glib2'` will load the `.so` file instead of the `.rb` file.

See this similar bug in `psych`: https://github.com/tenderlove/psych/issues/223.

**Comment by kou** [comment]
The problem may be a RubyGems problem. Could you try to use the latest RubyGems?ff

**Comment by mvz** [comment]
Yes, that fixed it. Thanks @kou. Skimming through the change log, it looks like it was fixed in https://github.com/rubygems/rubygems/pull/1137, which was released as part of rubygems 2.4.6.

**Comment by kou** [comment]
It will be fixed by using RubyGems 2.4.6 or later.
I close this.
